Refactor number of pages alias (fix #1090)

andersonhc opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Some guidelines are provided like this, in HTML comments, to expedite the code review before merging your contribution.

First, please make sure that you have read the documentation page on fpdf2 development:
https://py-pdf.github.io/fpdf2/Development.html

If you're new to contributing to open source projects,
you might find this free video course helpful: http://kcd.im/pull-request
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

e.g. Fixes #0 <!-- This will automatically close issue #0 once the PR is merged: https://help.github.com/en/articles/closing-issues-using-keywords -->

<!-- Have you done all of these things?  -->
**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [x] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [x] A mention of the change is present in `CHANGELOG.md`

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -279,7 +279,9 @@ def __init__(
         """
         self.page = 0  # current page number
-        self.pages = {}  # array of PDFPage objects starting at index 1
+        self.pages: Dict[int, PDFPage] = (
+            {}
+        )  # array of PDFPage objects starting at index 1
```
```suggestion
        # array of PDFPage objects starting at index 1:
        self.pages: Dict[int, PDFPage] = {}
```
Just a suggestion to make the code clearer

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -332,6 +333,25 @@ def render_pdf_text_core(self, frag_ws, current_ws):
+        return f"::alias:{self.uuid}::"
+
+    def render_pdf_text(self, *args, **kwargs):
+        self._render_args = args
+        self._render_kwargs = kwargs
```
Shouldn't those attributes be defined in the constructor?

andersonhc replied:
I added docstrings to explain this method

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -5158,8 +5175,14 @@ def output(
             # Generating .buffer based on .pages:
             if self._toc_placeholder:
                 self._insert_table_of_contents()
-            if self.str_alias_nb_pages:
```
Why removing this IF?

gmischler reviewed (approved):
Alias = "A different **_name_** for something"

That's not really what we're dealing with here, at all.

Yes, I know that `.str_alias_nb_pages` has had that name for like forever, but maybe we shouldn't propagate it to all the internal code dealing with text substitution.
In #1202 I tried to use terminology that describes better what really happens. I probably haven't found the absolute best wording, but it may help as a more reasonable starting point.

Using names that don't actually reflect the purpose of the named elements makes the code really confusing to read and hard to understand. I'm not a native english speaker, but I like to be as precise as possible about those things.

Other than that, I expect the result to "just work"... 😉

andersonhc commented:
@Lucas-C @gmischler 
This PR should be ready now.
I renamed the internal methods from "alias" to "text substitution" but didn't touch the ones that already existed (like alias_nb_pages) to keep backwards compatibility.

Lucas, I changed the text shaping enable/disable test you created recently because after this change the replaced number of pages is similar with text shaping on or off. I replaced by a text with ligatures that have internal and visual differences.
